src/include/migraphx/op/unique.hpp

```diff
@@ -0,0 +1,320 @@
+    // x_rev_indices: [0, 1, 1, 2, 3, 2] --- x seen in terms of indices of unique sequence y
+    // y_count: [1, 2, 2, 1] -- count at each y_index. sum = len(x)
+
+    // NOTE: y [2, 1, 3, 4]   --- the unique output is constructed from x[y_indices[...]]
+    // Output data structures: y_indices, x_rev_indices, y_count are processed inline.
+
+    template <class T>
+    auto unsorted_uniq_indices(const T& input_data, size_t chunk_sz) const
+    {
+        auto idx_less_fn = [&](auto idx1, auto idx2) {
+            return std::lexicographical_compare(input_data.begin() + idx1,
+                                                input_data.begin() + idx1 + chunk_sz,
+                                                input_data.begin() + idx2,
+                                                input_data.begin() + idx2 + chunk_sz);
+        };
+        std::map<size_t, size_t, decltype(idx_less_fn)> uniq_val_map(idx_less_fn);
+
+        std::tuple<std::vector<std::size_t>, std::vector<std::size_t>, std::vector<std::size_t>> rv;
+        auto& [y_indices, x_rev_indices, y_count] = rv;
+
+        // go through all the elements and add the unique elements into the map..
+        // inline processing for outputs: y_indices, x_rev_indices, y_count
+        size_t count_x = input_data.size();
+        for(size_t f_idx = 0; f_idx < count_x; f_idx += chunk_sz)
+        {
+            auto [itr, added_new] = uniq_val_map.insert({f_idx, y_indices.size()});
+            if(added_new)
+            {
+                y_count.push_back(0);
+                y_indices.push_back(x_rev_indices.size());
+            }
+            y_count[itr->second]++;
+            x_rev_indices.push_back(itr->second);
+        }
+
+        return rv;
+    }
+
+    // Axis. Default: none. Range: [-rank, rank-1]
+    std::optional<int64_t> axis;
+
+    // Sorted, Default: 1= sorted. 0 = unsorted.
+    bool sorted = true;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.axis, "axis"), f(self.sorted, "sorted"));
+    }
+
+    std::string name() const { return "unique"; }
+
+    shape compute_shape(std::vector<shape> inputs) const
+    {
+        check_shapes{inputs, *this}.has(1);
+
+        auto& sh_x         = inputs[0];
+        auto lens_x        = sh_x.lens();
+        size_t dim_x       = sh_x.ndim();
+        size_t max_uniq_ct = sh_x.elements();
+        std::vector<shape::dynamic_dimension> d_out;
+
+        // min = 1 unique element; max = full dimension along the axis
+        if(axis)
+        {
+            int64_t t_axis = migraphx::tune_axis(dim_x, *axis, name());
+            if(t_axis != 0)
+                MIGRAPHX_THROW("Unique: Only supports axis = 0 or None");
+
+            // only axis = 0 is supported:
+            max_uniq_ct /= lens_x[0];
+            d_out.push_back({1, max_uniq_ct});
+            for(size_t idx = 1; idx < dim_x; idx++)
+                d_out.push_back({lens_x[idx], lens_x[idx]});
```

Comment:
I think this would be simpler to use `to_dynamic` here:

```cpp
d_out = sh_x.to_dynamic().dyn_dims();
d_out.front().min = 1;
d_out.front().max = max_uniq_ct;
```

Reply:
Logically speaking, one shouldn't change a shape once defined. That is the reason it was done this way..

Reply:
I shouldn't worry about that too much.  Shapes are disposable, lightweight objects, and the compute_shape() method is definitely the most appropriate place to do any wrangling of shapes it takes to create the desired output.  OTOH, either code version looks fine to me.

Reply:
Thanks. We should change shapes  within its implementation in shape.cpp/hpp.

Reply:
> Logically speaking, one shouldn't change a shape once defined. That is the reason it was done this way..

This is not mutating the shape object(which is not even possible since all the methods to `shape` have to be `const`). Instead, this is just modifying `dout`, which is just a vector of dynamic dimensions.

Reply:
Well firstly, a to_dynamic() is used, that is why I didn't want to mess with the results of it. Finally, I am not sure what is simpler here to use the suggested 3 lines for one line anyways.. -- and perhaps repeated for two sets of logic each.. In any case, are you proposing a value add of some kind here, or can it be left as it is otherwise. Thanks.

Reply:
> Well firstly, a to_dynamic() is used, that is why I didn't want to mess with the results of it.

Why? I am not sure I follow this. 

> Finally, I am not sure what is simpler here to use the suggested 3 lines for one line anyways..

To be clear, the suggestion is to replace the above 3 lines not the last line. And you could probably make it 2 lines(if you are concerned about the number of lines) if you write it like:

```cpp
d_out = sh_x.to_dynamic().dyn_dims();
d_out.front() = shape::dynamic_dimension{1, max_uniq_ct};
```

So this suggestion is even shorter than what you have written. 

> and perhaps repeated for two sets of logic each..

Are you referring to the non-axis branch? I am not suggesting a change to the non-axis branch, so there is no repeat of the logic. 

>  In any case, are you proposing a value add of some kind here

This is to make it clear you are using the same dimensions from `x` for everything except the first dimension. Which also matches the pattern I suggested [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/2072#discussion_r1400960751) except its using dynamic dimensions instead of integers.

Reply:
Messy use of to_dynamic() and to its dyn_dims(), and then an effort to modify it.  Instead of using the api to do dynamic shapes. Net value add  is negative, IMO.  However, it is done.

Reply:
Thanks, this will be much easier to tweak to support a different axis in the future as well(we just replace the `0` with `*axis`).